### ISSUE TITLE
Event loop blocked by >3s

### ISSUE THREAD

**Issue by dapplion** [ChainSafe/lodestar#1990]
Two recent syncs revealed event loop lag over 3s multiple times. We should investigate what action is causing so much lag and control it.

Note that event loop lag correlates with peer loss

#### Sync 1

![Screenshot from 2021-01-24 08-05-07](https://user-images.githubusercontent.com/35266934/105752056-1e43ab00-5f47-11eb-8940-6c432a6b6c31.png)

#### Sync 2

![Screenshot from 2021-01-23 10-58-17](https://user-images.githubusercontent.com/35266934/105752068-226fc880-5f47-11eb-896f-de27cc7c1371.png)

**Comment by twoeths** [comment]
sometimes I see epoch processing increased significantly and then came back. This is from my current pyrmont sync
```
@@@ epochTransition CachedBeaconState slot 90240 time 1255 average 1247.4973646382361
@@@ epochTransition CachedBeaconState slot 90272 time 2387 average 1248.0431034482754
@@@ epochTransition CachedBeaconState slot 90304 time 6620 average 1250.6146481570124
@@@ epochTransition CachedBeaconState slot 90336 time 6828 average 1253.2832535885163
@@@ epochTransition CachedBeaconState slot 90368 time 7466 average 1256.2544237207073
@@@ epochTransition CachedBeaconState slot 90464 time 5795 average 1258.4239961759079
@@@ epochTransition CachedBeaconState slot 90496 time 6486 average 1260.921643573817
@@@ epochTransition CachedBeaconState slot 90528 time 5762 average 1263.071155682903
@@@ epochTransition CachedBeaconState slot 90560 time 6552 average 1265.5957040572787
@@@ epochTransition CachedBeaconState slot 90592 time 5712 average 1267.7170801526713
@@@ epochTransition CachedBeaconState slot 90624 time 6269 average 1270.102050548402
@@@ epochTransition CachedBeaconState slot 90656 time 6701 average 1272.6906577693037
@@@ epochTransition CachedBeaconState slot 90688 time 8099 average 1275.9428299190085
@@@ epochTransition CachedBeaconState slot 90720 time 7594 average 1278.951428571428
@@@ epochTransition CachedBeaconState slot 90752 time 6840 average 1281.5982865302233
@@@ epochTransition CachedBeaconState slot 90784 time 7429 average 1284.5228353948617
@@@ epochTransition CachedBeaconState slot 90816 time 6222 average 1286.8706609605322
@@@ epochTransition CachedBeaconState slot 90848 time 6130 average 1289.1725285171099
@@@ epochTransition CachedBeaconState slot 90880 time 5094 average 1290.9800475059378
@@@ epochTransition CachedBeaconState slot 90912 time 7397 average 1293.8793922127252
@@@ epochTransition CachedBeaconState slot 90944 time 5883 average 1296.0574276222112
@@@ epochTransition CachedBeaconState slot 90976 time 10526 average 1300.435958254269
@@@ epochTransition CachedBeaconState slot 91008 time 5963 average 1302.6467520151725
@@@ epochTransition CachedBeaconState slot 91040 time 1430 average 1302.7071090047389
```

average epoch transition is ~1.3s while sometimes it increased significantly to more than 10s!

**Comment by dapplion** [comment]
Is there something we can split out the tasks so it doesn't take so long at once?

**Comment by twoeths** [comment]
I'll try to regenerate respective BeaconStates to reproduce on local environment to see if we can improve it.

**Comment by twoeths** [comment]
I don't see a slow epoch transition after I regenerate state and run state transition through block 90976, it's only around 1.1s.

So I think it's just my laptop was too busy at that time since it happened at adjacent epochs. We need to make a new metric for epoch transition and follow it once we deploy to a server.

I'm attaching the state and block for reference.
[archivedstate_90975.zip](https://github.com/ChainSafe/lodestar/files/5878132/archivedstate_90975.zip)
[archivedblock_90976.zip](https://github.com/ChainSafe/lodestar/files/5878138/archivedblock_90976.zip)

**Comment by dapplion** [comment]
Thanks for profiling it! Even tho it's 1.1s, could you investigate the percentage CPU spent in each major routine?

**Comment by twoeths** [comment]
here's the statistic based on https://github.com/ChainSafe/lodestar/blob/master/packages/lodestar-beacon-state-transition/test/perf/epoch_processing.test.ts and the attached state 90975
```
2021-01-27 14:07:42 []                 info: prepareEpochProcessState  - duration=323ms
    1) prepareEpochProcessState
2021-01-27 14:07:42 []                 info: processJustificationAndFinalization  - duration=1ms
    ✓ processJustificationAndFinalization
2021-01-27 14:07:42 []                 info: processRewardsAndPenalties  - duration=498ms
    ✓ processRewardsAndPenalties (498ms)
2021-01-27 14:07:42 []                 info: processRegistryUpdates
    ✓ processRegistryUpdates
2021-01-27 14:07:42 []                 info: processSlashings  - duration=23ms
    ✓ processSlashings
2021-01-27 14:07:42 []                 info: processFinalUpdates  - duration=21ms
    ✓ processFinalUpdates
2021-01-27 14:07:42 []                 info: processForkChanged
    ✓ processForkChanged
2021-01-27 14:07:42 []                 info: rotateEpoch  - duration=165ms
```

and for block processing

```
2021-01-27 14:15:56 []                 info: Process block 90976  - duration=164ms
```

**Comment by twoeths** [comment]
it's worth to look into this https://developer.mozilla.org/en-US/docs/Web/API/HTML_DOM_API/Microtask_guide

currently we use `async/await` across the code base, in a lot of scenarios it keeps adding items to `microtasks` queue in the same event loop iteration hence that iteration takes time.

**Comment by twoeths** [comment]
sometimes it takes almost 1s just for the `readOnlyMap` inside `processRewardsAndPenalties`, I suspect it much slower in our server
<img width="1144" alt="Screen Shot 2021-02-08 at 14 52 36" src="https://user-images.githubusercontent.com/10568965/107190536-7d142480-6a1d-11eb-87f7-a03c57bbb961.png">

Most of the time it took only 200ms-300ms for that, also I notice that there are a lot of `gc` calls there.

One more thing to take a look is the BigInt calculations
<img width="887" alt="Screen Shot 2021-02-08 at 15 00 35" src="https://user-images.githubusercontent.com/10568965/107191251-7a65ff00-6a1e-11eb-8582-fdb3d0bc1a6a.png">

**Comment by dapplion** [comment]
Damn.. Thanks @tuyennhv for the analysis! Where is the file `uint.ts` from? Our ssz library?

EDIT: Yes it's ssz, src/types/basic/uint.ts

EDIT2: @tuyennhv The code from your debugger is different than master https://github.com/ChainSafe/ssz/blob/7639b2806713f6f352b43424a5aadca0da1779e2/src/types/basic/uint.ts#L70 , is that expected?

**Comment by twoeths** [comment]
> Damn.. Thanks @tuyennhv for the analysis! Where is the file `uint.ts` from? Our ssz library?
> 
> EDIT: Yes it's ssz, src/types/basic/uint.ts
> 
> EDIT2: @tuyennhv The code from your debugger is different than master https://github.com/ChainSafe/ssz/blob/7639b2806713f6f352b43424a5aadca0da1779e2/src/types/basic/uint.ts#L70 , is that expected?

there are NumberUintType and BigIntUintType in that same file, we need to improve both by limiting the use of BigInt if possible. Finally I was able to reproduce the issue in the performance test
```ts
it.only("reproduce readOnlyMap issue", async function () {
    this.timeout(0);
    let minTime = Number.MAX_SAFE_INTEGER;
    let maxTime = 0;
    const originalState = state.getOriginalState();
    for (let i = 0; i < 20000; i++) {
      const start = Date.now();
      readOnlyMap(originalState.balances, (balance) => balance);
      const duration = Date.now() - start;
      if (duration < minTime) minTime = duration;
      if (duration > maxTime) maxTime = duration;
    }
    console.log("@@@ minTime", minTime, "maxTime", maxTime);
  });
```
this prints out
```
2021-02-09 08:15:42 []                 info: Loaded state slot=756416, numValidators=114038
@@@ minTime 140 maxTime 1109
```

**Comment by twoeths** [comment]
there are different improvements implemented, this should be tracked in #2046
